### PR TITLE
Serialize enum values as String

### LINKED ISSUE #885: Serialize enums as a string instead of number
### Discussed in https://github.com/serverpod/serverpod/discussions/879

<div type='discussions-op-text'>

<sup>Originally posted by **andriyndev** May  6, 2023</sup>
(Copy-pasted from https://github.com/serverpod/serverpod/discussions/715 because it's a separate issue)

I think enum's values should never be serialized by numeric values because if the app's developer edits the enum so that the values of the current elements are shifted, it'll cause nasty consequences. For example, user selects one value, but it's written to the database as another one, and the worse is that they'll see it as they specified it while other users with more recent versions of the app will see it as it was written to the database. It might be especially dangerous if the data represents delicate information like user's religion, political view, sexual orientation, gender etc. And even if the app's developers are aware of this nuance, they still might be susceptible to human factor. And if the mistake will be made, it might be impossible to fix the consequences without interacting with each user who triggered writes of the data to the database because there is no way for the app's developers to know whether the value is what the user actually selected.</div>

### PR DESCRIPTION
**Pull request by lukehutch**
Obsoletes #1080.
Fixes #885.

Allow enums to be serialized as strings rather than integers, for robustness.

Rebased on latest git.
Complete with tests.

I want to enable this by default for 1.2, but some of you said that should wait until 2.0, so the option is set to false by default.

I hope you can please merge this soon, because this is the most annoying PR to have to keep in sync with upstream, since the generated code keeps causing merge conflicts with the generated code you have checked into git. Thanks!

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

### CODE REVIEW

**Comment by SandPod** [comment]
Hey Luke!

We have capacity to make a push with this change right now. If it is OK with you I could take over and do the finalizing fixes to get it in?

I will be pushing any changes directly to this branch.

File: packages/serverpod/lib/src/database/value_encoder.dart [hunk 1]
```diff
@@ -20,6 +20,22 @@ class ValueEncoder extends PostgresTextEncoder {
           escapeStrings: escapeStrings);
     } else if (input is UuidValue) {
       return "'${input.uuid}'";
+    } else if (input is Enum) {
```
Comment (SandPod) [inline packages/serverpod/lib/src/database/value_encoder.dart:23]: Chore: Add an e2e test for both enum types.

This is a tricky one, but we should test our golden path for serialization of both enum types to the client to ensure that everything is working as expected.

File: packages/serverpod/lib/src/database/value_encoder.dart [hunk 2]
```diff
@@ -20,6 +20,22 @@ class ValueEncoder extends PostgresTextEncoder {
...
+      // Determine whether the Enum is serialized as an integer or as a String
+      var enumValueAsJson = (input as dynamic)?.toJson();
+      if (enumValueAsJson is int) {
```
Comment (SandPod) [inline packages/serverpod/lib/src/database/value_encoder.dart:26]: Question: For my knowledge - What happens in the case that the enum string is a series of numbers. For example if we have a string that is `"12345"`. Would this parse it correctly? Is that the reson you use the `toJson()` over `int.parse(...)`?

Comment (SandPod) [reply packages/serverpod/lib/src/database/value_encoder.dart:26]: (It might not be a real case since enum values can't be numbers, but I'm just curious.)

File: packages/serverpod/test/database/columns/column_enum_test.dart [hunk 3]
```diff
@@ -8,80 +8,159 @@ enum TestEnum {
 }
 
 void main() {
-  group('Given a ColumnEnum', () {
+  group('Given a ColumnEnumSerializedAsString', () {
```
Comment (SandPod) [inline packages/serverpod/test/database/columns/column_enum_test.dart:11]: Chore: These tests should be separated into two test files named after the enum they are testing.

File: tools/serverpod_cli/test/util/protocol_helper_test.dart [hunk 4]
```diff
@@ -18,20 +18,23 @@ GeneratorConfig createGeneratorConfig([
...
+final serverRootDir = Directory(join(
+  'test',
+  'util',
+  'test_assets',
+  'protocol_helper',
+  'has_serverpod_server_project',
+  'test_server',
+));
+
+final config = createGeneratorConfig(split(serverRootDir.path));
```
Comment (SandPod) [inline tools/serverpod_cli/test/util/protocol_helper_test.dart:34]: Suggestion: Move inside of test `main` method to scope their lifetime.

These variables are defined as soon as this files is included somewhere. Since assets created in a test should only be local to the tests the intention is more clear if we scope these to the main method of the test. We can still share them between the groups.

File: tools/serverpod_cli/test/generator/dart/server_code_generator/enum_field_test.dart [hunk 5]
```diff
@@ -40,12 +39,19 @@ void main() {
     });
 
     test('then generated enum has static fromJson method', () {
-      expect(codeMap[expectedFileName],
-          contains('static Example? fromJson(int index)'));
+      expect(
+          codeMap[expectedFileName],
+          contains(config.serializeEnumValuesAsStrings
+              ? 'static Example? fromJson(String name)'
+              : 'static Example? fromJson(int index)'));
```
Comment (SandPod) [inline tools/serverpod_cli/test/generator/dart/server_code_generator/enum_field_test.dart:46]: Chore: Create separate groups for both enum types used in the test and only validate one in each group.

Depending on what the default value is in our GeneratorConfigBuilder we won't exercise one of these paths. We want to ensure that we support all paths possible to configure and should therefore have a test validating it.

File: tools/serverpod_cli/test/database/extentions/pgsql_code_generation_test.dart [hunk 6]
```diff
@@ -2,9 +2,13 @@ import 'package:recase/recase.dart';
 import 'package:serverpod_cli/analyzer.dart';
 import 'package:serverpod_cli/src/database/create_definition.dart';
 import 'package:serverpod_cli/src/test_util/builders/class_definition_builder.dart';
+import 'package:serverpod_cli/src/test_util/builders/generator_config_builder.dart';
 import 'package:test/test.dart';
 
 void main() {
+  // Cache GeneratorConfig
+  GeneratorConfigBuilder().withName('database_action_definition_test').build();
```
Comment (SandPod) [inline tools/serverpod_cli/test/database/extentions/pgsql_code_generation_test.dart:10]: Question: For my knowledge - Why is this required? And what do we gain by doing this?

File: tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart [hunk 7]
```diff
@@ -2,11 +2,15 @@ import 'package:serverpod_cli/analyzer.dart';
...
 import 'package:serverpod_cli/src/test_util/builders/class_definition_builder.dart';
 import 'package:serverpod_cli/src/test_util/builders/foreign_relation_definition_builder.dart';
+import 'package:serverpod_cli/src/test_util/builders/generator_config_builder.dart';
 import 'package:serverpod_cli/src/test_util/builders/serializable_entity_field_definition_builder.dart';
 import 'package:serverpod_service_client/serverpod_service_client.dart';
 import 'package:test/test.dart';
 
 void main() {
+  // Cache GeneratorConfig
+  GeneratorConfigBuilder().withName('database_action_definition_test').build();
```
Comment (SandPod) [inline tools/serverpod_cli/test/database/create_definition/database_action_definition_test.dart:12]: Question: For my knowledge - Why is this required? And what do we gain by doing this?

File: tools/serverpod_cli/lib/src/config/config.dart [hunk 8]
```diff
@@ -34,7 +34,21 @@ class GeneratorConfig {
     required List<String> relativeDartClientPackagePathParts,
     required this.modules,
     required this.extraClasses,
-  }) : _relativeDartClientPackagePathParts = relativeDartClientPackagePathParts;
+    required this.serializeEnumValuesAsStrings,
+  }) : _relativeDartClientPackagePathParts =
+            relativeDartClientPackagePathParts {
+    assert(_instance == null, 'GeneratorConfig should only be loaded once.');
+    _instance = this;
+  }
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/config/config.dart:42]: Question: This is a good change but has this branch been tested with all different commands we support in the CLI? I think we for example in the `create-migration` command actually load the generator config multiple times.

This might be a change for a different PR where we fix all these issues and add this to prevent more cases from popping up.

File: tools/serverpod_cli/lib/src/config/config.dart [hunk 9]
```diff
@@ -34,7 +34,21 @@ class GeneratorConfig {
...
+
+  /// The current [GeneratorConfig] instance.
+  static GeneratorConfig? _instance;
+
+  /// Get the current [GeneratorConfig] instance.
+  static GeneratorConfig get instance {
+    assert(_instance != null, 'GeneratorConfig is not loaded.');
+    return _instance!;
+  }
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/config/config.dart:51]: Note: I'm not sure about the singleton approach for the generator config. My hunch tells me it should rather live in one place than be confined as a singleton.

But this might be the right approach since it is accessed on different abstraction levels in multiple places in the code right now.

File: tools/serverpod_cli/lib/src/config/config.dart [hunk 10]
```diff
@@ -214,6 +231,20 @@ class GeneratorConfig {
...
+        serializeEnumValuesAsStringsConfig is! bool) {
+      throw const FormatException(
+          'Option "serializeEnumValuesAsStringsConfig" must be of type bool '
+          'in config/generator.yaml');
+    }
+    var serializeEnumValuesAsStrings =
+        // TODO: switch the default from `false` to `true` in Serverpod 2.0.
+        // Also switch the default in:
+        // tools/serverpod_cli/lib/src/test_util/builders/generator_config_builder.dart
+        (serializeEnumValuesAsStringsConfig as bool?) ?? false;
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/config/config.dart:246]: Awesome 🎉

Comment (SandPod) [inline tools/serverpod_cli/lib/src/config/config.dart:246]: Suggestion: add tests for this.

Since this ensure that we don't introduce a breaking change I think we should have a tests that validates that this works as expected.

File: tools/serverpod_cli/lib/src/test_util/builders/generator_config_builder.dart [hunk 11]
```diff
@@ -74,6 +79,12 @@ class GeneratorConfigBuilder {
     return this;
   }
 
+  GeneratorConfigBuilder serializeEnumsValuesAsStrings(
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/test_util/builders/generator_config_builder.dart:82]: Suggestion: Rename to `withEnumsSerializedAsStrings`.

Following the naming convention in the file this feels like a more matching name. Then it doesn't need to take a bool either.

**Review by SandPod — CHANGES_REQUESTED** [review]
Hey!

Let's make a push to take this in.

There are a couple of changes required to assert that we don't introduce any breaking changes.

**Comment by lukehutch** [comment]
@SandPod So I guess I'll wait for your responses to the foregoing questions first, before I work on the suggestions, because maybe a lot of this code can go away if we put a proper migration in place. I think with a proper migration, doing this for 1.2 would be reasonable. Do you think?

File: packages/serverpod/lib/src/database/value_encoder.dart (hunk 2 above)
Comment (lukehutch) [reply packages/serverpod/lib/src/database/value_encoder.dart:26]: I copied that code snippet `(input as dynamic)?.toJson()` from somewhere else in Serverpod when I first wrote this code!

This is converting an int or string to a string though. `int.parse` would go the other direction.

File: tools/serverpod_cli/lib/src/config/config.dart (hunk 8 above)
Comment (lukehutch) [reply tools/serverpod_cli/lib/src/config/config.dart:42]: The question you raise is a direct result of this particular change being a very ugly hack! See my longer comment in the discussion. I would like to see this go away by entirely getting rid of serialization to integers.

File: tools/serverpod_cli/lib/src/config/config.dart (hunk 9 above)
Comment (lukehutch) [reply tools/serverpod_cli/lib/src/config/config.dart:51]: Totally agree, I hate this part of the PR. The alternative was to pass the config value all over the source tree. I tried doing this and ended up modifying dozens and dozens of files to get the value where I needed it to be. It ended up atrocious.

**Comment by lukehutch** [comment]
@SandPod thanks for the clarifications, I agree this is the way to move forwards. I wouldn't attempt changes at the scale you are suggesting, because I'm way too busy working on my own project. I'll leave this one in your hands, and you can close this PR when something better exists!

**Comment by lukehutch** [comment]
How about we do implement this using only the `EnumDefinition` and codegen? And then the default serialization type (if one is not specified in an enum protocol file) can simply be int until 2.0, then string thereafter.

(All the other `GeneratorConfig` changes you are proposing could be done separately as a broader refactoring process. But I'm starting to think there shouldn't be a default serialization format specified in the `GeneratorConfig`.)

The main thing that would have to be looked at is how to get the serialization type through to `types.dart`, methods `databaseType` and `columnType`. (The `value_encoder.dart` changes in my PR would still work, regardless of whether `toJson` returns an `int` or a `String`.)

**Comment by lukehutch** [comment]
@SandPod I make a quick attempt at this new approach in #1521, please check it... Thanks!

**Comment by SandPod** [comment]
@lukehutch - Understandable, I'll have a look at your new approach. I sounds like it could be a valid approach to first implement the "lowest level" of configurability in protocol files, then we could later on intrude a global default.

**Comment by lukehutch** [comment]
Closing this in favor of #1521.
